# Consider the game of tennis when *deuce* is reached

# If a player wins the next point, he has *advantage*

# On the following point, he either wins the game or the game returns to *deuce*

# Assume that for any point, player A has probability 0.6 of winning the point and player B has probability 0.4 of winning the point

## a) Set this up as a Markov chain with states: A wins, B wins, advantage A, deuce, advantage B

## b) Find the absorption probabilities

## c) At deuce, find the expected duration of the game and the probability that B will win

_________

# a)

## We know that A wins and B wins are absorbing states, so we $P$'s canonical form is given by:

# $P = \begin{pmatrix}0 & 0.6 & 0.4 & 0 & 0\\ 0.4 & 0 & 0 & 0.6 & 0\\ 0.6 & 0 & 0 & 0 & 0.4\\ 0 & 0 & 0 & 1 & 0\\ 0 & 0 & 0 & 0 & 1\end{pmatrix}$

# where the rows/columns states are deuce, Adv. A, Adv. B, Win A, and Win B respectively

____

# b)

# $\implies Q = \begin{pmatrix}0 & 0.6 & 0.4\\ 0.4 & 0 & 0\\ 0.6 & 0 & 0\end{pmatrix}$

# $\implies (I-Q) = \begin{pmatrix}1 & -0.6 & -0.4\\ -0.4 & 1 & 0\\ -0.6 & 0 & 1\end{pmatrix}$

**Using numpy to calculate inverse**

In [1]:
import numpy as np

In [2]:
matrix = np.array([[1, -0.6, -0.4],[-0.4, 1, 0], [-0.6, 0, 1]])

In [3]:
N = np.linalg.inv(matrix)

In [4]:
N

array([[ 1.92307692,  1.15384615,  0.76923077],
       [ 0.76923077,  1.46153846,  0.30769231],
       [ 1.15384615,  0.69230769,  1.46153846]])

# $\implies N = \begin{pmatrix}1.92307692 & 1.15384615 & 0.76923077\\ 0.76923077 & 1.46153846 & 0.30769231\\ 1.15384615 & 0.69230769 & 1.46153846\end{pmatrix}$

# $R = \begin{pmatrix}0 & 0\\ 0.6 & 0\\ 0 & 0.4\end{pmatrix}$

In [5]:
R = np.array([[0,0],[0.6,0],[0,0.4]])

In [6]:
NR = np.matmul(N,R)

In [7]:
NR

array([[ 0.69230769,  0.30769231],
       [ 0.87692308,  0.12307692],
       [ 0.41538462,  0.58461538]])

# $\implies NR = B = \begin{pmatrix}0.69230769 & 0.30769231\\ 0.87692308 & 0.12307692\\ 0.41538462 & 0.58461538\end{pmatrix}$

_____

# c)

# Since we already calculated $B$, we look at top right value in the matrix to see the probability that $B$ wins given a deuce

# $P(\text{B wins}|\text{Deuce}) = 0.307$

# We find the time to absorption by calculating $Nc$ and taking the value in the top row

In [8]:
c = np.array([[1],[1],[1]])
Nc = np.matmul(N,c)

In [9]:
Nc

array([[ 3.84615385],
       [ 2.53846154],
       [ 3.30769231]])

# $\implies t = Nc = \begin{pmatrix}3.84615385\\ 2.53846154\\ 3.30769231\end{pmatrix}$

# Therefore, if the game has reached a deuce, we expect it to end in 3.85 rallies

______

### Let's confirm this result by simulation

In [25]:
def game_sim():
    A_score = 0
    B_score = 0
    
    i = 0
    while (A_score<2)&(B_score<2):
        i += 1
        rally = np.random.random()
        if rally > 0.4:
            if B_score == 1:
                B_score = 0
            else:
                A_score += 1
        else:
            if A_score == 1:
                A_score = 0
            else:
                B_score += 1
                
    return 1*(A_score<B_score), i

In [26]:
list_wins = []
list_rallies = []

for j in range(10000):
    win, rallies = game_sim()
    list_wins.append(win)
    list_rallies.append(rallies)

In [27]:
np.mean(list_wins)

0.30769999999999997

In [28]:
np.mean(list_rallies)

3.8490000000000002

### Both are bang on